plugins/catalog-backend-module-gerrit/src/providers/GerritEntityProvider.ts

```diff
@@ -0,0 +1,165 @@
+          `No gerrit integration found that matches host ${providerConfig.host}`,
+        );
+      }
+      providers.push(
+        new GerritEntityProvider(
+          providerConfig,
+          integration,
+          options.logger,
+          options.schedule,
+        ),
+      );
+    });
+    return providers;
+  }
+
+  constructor(
+    config: GerritProviderConfig,
+    integration: GerritIntegration,
+    logger: Logger,
+    schedule: TaskRunner,
+  ) {
+    this.config = config;
+    this.integration = integration;
+    this.logger = logger.child({
+      target: this.getProviderName(),
+    });
+    this.scheduleFn = this.createScheduleFn(schedule);
+  }
+
+  getProviderName(): string {
+    return `gerrit-provider:${this.config.id}`;
+  }
+
+  async connect(connection: EntityProviderConnection): Promise<void> {
+    this.connection = connection;
+    await this.scheduleFn?.();
+  }
+
+  private createScheduleFn(schedule: TaskRunner): () => Promise<void> {
+    return async () => {
+      const taskId = `${this.getProviderName()}:refresh`;
+      return schedule.run({
+        id: taskId,
+        fn: async () => {
+          const logger = this.logger.child({
+            class: GerritEntityProvider.prototype.constructor.name,
+            taskId,
+            taskInstanceId: uuid.v4(),
+          });
+
+          try {
+            await this.refresh(logger);
+          } catch (error) {
+            logger.error(error);
+          }
+        },
+      });
+    };
+  }
+
+  async refresh(logger: Logger): Promise<void> {
+    if (!this.connection) {
+      throw new Error('Gerrit discovery connection not initialized');
+    }
+
+    let response: Response;
+
+    const baseProjectApiUrl = getGerritProjectsApiUrl(this.integration.config);
+    const projectQueryUrl = `${baseProjectApiUrl}?${this.config.query}`;
+    try {
+      response = await fetch(projectQueryUrl, {
+        method: 'GET',
+        ...getGerritRequestOptions(this.integration.config),
+      });
+    } catch (e) {
+      throw new Error(
+        `Failed to list Gerrit projects for query ${this.config.query}, ${e}`,
+      );
+    }
+    const gerritProjectsResponse = (await parseGerritJsonResponse(
+      response as any,
+    )) as GerritProjectQueryResult;
+    const projects = Object.keys(gerritProjectsResponse);
+
+    const locations = projects.map(project => this.createLocationSpec(project));
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: locations.map(location => ({
+        locationKey: this.getProviderName(),
+        entity: locationSpecToLocationEntity({ location }),
+      })),
+    });
+    logger.info(`Found ${locations.length} locations.`);
+  }
+
+  createLocationSpec(project: string): LocationSpec {
+    return {
+      type: 'url',
+      target: `${this.integration.config.gitilesBaseUrl}/${project}/+/refs/heads/${this.config.branch}/catalog-info.yaml`,
+      presence: 'optional',
```

Comment:
I assume all these locations are still only potential locations and you did not verify that the catalog-info.yaml files actually exists, right?

Reply:
Yes, it is up to the "UrlReader" to check if the location is valid.